### DISCUSSION TITLE
How do I go about implementing an image reconstruction method (inverse solver) using the results from meep?

### DISCUSSION THREAD

**Discussion by cyasing** [NanoComp/meep#2427] [Q&A] [answered]
I need to implement an image reconstruction algorithm using the electromagnetic scattering data generated from meep (ideally the Distorted Born Iterative Method). Is there any support for meep in this? I checked adjoint solver, but it does not seem to be suited. If not, do I make a separate program, using libraries like Eigen, for example? Also, is there anything I should know about handling complex field values?

**Comment by stevengj** [comment, marked as answer]
As I understand it, DBIM is essentially a type of volume-integral method (VIM) that you solve iteratively while updating the reconstruction, which is a totally different solver technique than Meep. So I'm a little confused about how you think DBIM could be applied here.

In principle you could use the adjoint solver for fullwave inversion (FWI), where the material-grid design variables would be your unknown density (your reconstruction), and your objective function would be the difference between the scattered field (e.g. via near2far) and your measured field (usually plus some regularization or prior).

**Reply by cyasing** [reply]
Yes, DBIM is indeed a different solver than what FDTD. I wanted to use MEEP at the comparison stage at each iteration, where it compares scattered data from the original object compared to that of the reconstructed one to improve reconstruction. Hence I wanted to know if there is any in-built way for it, instead of having to implement DBIM separately.

Looks like the adjoint solver is indeed workable to get results to begin with, but DBIM will need a separate implementation.
